File: src/services/translationService.js
/**
 * Translation Service
 * Handles communication with LLM providers for text translation
 */

import { invoke } from '@tauri-apps/api/core'
import { isTauriEnv } from '../utils/environment.js'

/**
 * Translation Service class for handling AI-powered translations
 */
export class TranslationService {
    constructor() {
        this.currentTranslation = null
        this.isTranslating = false
    }

    /**
     * Translate text to multiple target languages
     * @param {string} inputText - Text to translate
     * @param {Array} targetLanguages - Array of target language names
     * @param {Object} config - LLM configuration
     * @returns {Promise<Object>} Translation result object
     */
    async translateText(inputText, targetLanguages, config) {
        if (!isTauriEnv()) {
            throw new Error('Translation requires Tauri environment. Please run: npm run dev')
        }

        if (!inputText || inputText.trim().length === 0) {
            throw new Error('Please enter text to translate')
        }

        if (!targetLanguages || targetLanguages.length === 0) {
            throw new Error('Please select at least one target language')
        }

        if (!config.serverUrl || !config.selectedModel) {
            throw new Error('Please configure LLM settings first')
        }

        this.isTranslating = true

        try {
            console.log('Starting translation request:', {
                text: inputText,
                targetLanguages,
                config: {
                    provider: config.provider,
                    serverUrl: config.serverUrl,
                    model: config.selectedModel
                }
            })

            const translationRequest = {
                text: inputText.trim(),
                target_languages: targetLanguages,
                llm_provider: config.provider,
                server_url: config.serverUrl,
                api_key: config.apiKey || '',
                model: config.selectedModel
            }

            const result = await invoke('translate_text', translationRequest)

            this.currentTranslation = {
                original: inputText.trim(),
                translations: result,
                timestamp: new Date().toISOString(),
                languages: targetLanguages,
                config: { ...config }
            }

            console.log('Translation completed successfully:', this.currentTranslation)
            return this.currentTranslation

        } catch (error) {
            console.error('Translation failed:', error)

            // Provide more specific error messages
            const errorMsg = error.message || error.toString()
            if (errorMsg.includes('Connection refused') || errorMsg.includes('network')) {
                throw new Error(`Cannot connect to ${config.provider} server at ${config.serverUrl}. Please check if the server is running and accessible.`)
            } else if (errorMsg.includes('Unauthorized') || errorMsg.includes('401')) {
                throw new Error('Authentication failed. Please check your API key.')
            } else if (errorMsg.includes('Not found') || errorMsg.includes('404')) {
                throw new Error(`Model "${config.selectedModel}" not found. Please check if the model is available.`)
            } else if (errorMsg.includes('timeout')) {
                throw new Error('Translation request timed out. The server may be busy.')
            } else {
                throw new Error(`Translation failed: ${errorMsg}`)
            }
        } finally {
            this.isTranslating = false
        }
    }

    /**
     * Save translation to file
     * @param {string} filename - Filename for the saved translation
     * @param {string} format - File format ('txt', 'json', 'md')
     * @returns {Promise<string>} Path to saved file
     */
    async saveTranslation(filename, format = 'txt') {
        if (!this.currentTranslation) {
            throw new Error('No translation available to save')
        }

        if (!isTauriEnv()) {
            throw new Error('File saving requires Tauri environment. Please run: npm run dev')
        }

        try {
            let content = ''

            switch (format.toLowerCase()) {
                case 'json':
                    content = JSON.stringify(this.currentTranslation, null, 2)
                    break

                case 'md':
                case 'markdown':
                    content = this.formatAsMarkdown(this.currentTranslation)
                    break

                case 'txt':
                default:
                    content = this.formatAsText(this.currentTranslation)
                    break
            }

            const savedPath = await invoke('save_translation_file', {
                filename: filename,
                content: content
            })

            console.log('Translation saved successfully:', savedPath)
            return savedPath

        } catch (error) {
            console.error('Failed to save translation:', error)
            throw new Error(`Failed to save translation: ${error.message || error}`)
        }
    }

    /**
     * Format translation as plain text
     * @param {Object} translation - Translation object
     * @returns {string} Formatted text content
     */
    formatAsText(translation) {
        let content = `Translation Results\\n`
        content += `Generated: ${new Date(translation.timestamp).toLocaleString()}\\n`
        content += `Original Language: Auto-detected\\n`
        content += `Target Languages: ${translation.languages.join(', ')}\\n`
        content += `Model: ${translation.config.selectedModel}\\n\\n`

        content += `Original Text:\\n${translation.original}\\n\\n`

        content += `Translations:\\n`
        for (const [lang, translatedText] of Object.entries(translation.translations)) {
            content += `\\n${lang}:\\n${translatedText}\\n`
        }

        return content
    }

    /**
     * Format translation as Markdown
     * @param {Object} translation - Translation object
     * @returns {string} Formatted Markdown content
     */
    formatAsMarkdown(translation) {
        let content = `# Translation Results\\n\\n`
        content += `**Generated:** ${new Date(translation.timestamp).toLocaleString()}  \\n`
        content += `**Original Language:** Auto-detected  \\n`
        content += `**Target Languages:** ${translation.languages.join(', ')}  \\n`
        content += `**Model:** ${translation.config.selectedModel}\\n\\n`

        content += `## Original Text\\n\\n`
        content += `${translation.original}\\n\\n`

        content += `## Translations\\n\\n`
        for (const [lang, translatedText] of Object.entries(translation.translations)) {
            content += `### ${lang}\\n\\n${translatedText}\\n\\n`
        }

        return content
    }

    /**
     * Get current translation
     * @returns {Object|null} Current translation object or null
     */
    getCurrentTranslation() {
        return this.currentTranslation
    }

    /**
     * Clear current translation
     */
    clearTranslation() {
        this.currentTranslation = null
    }

    /**
     * Get translation state
     * @returns {Object} Translation state
     */
    getTranslationState() {
        return {
            isTranslating: this.isTranslating,
            hasTranslation: !!this.currentTranslation,
            currentTranslation: this.currentTranslation
        }
    }

    /**
     * Format translation for display
     * @param {Object} translation - Translation object
     * @returns {Object} Formatted translation for UI display
     */
    formatForDisplay(translation = null) {
        const trans = translation || this.currentTranslation
        if (!trans) return null

        return {
            original: trans.original,
            translations: trans.translations,
            languages: trans.languages,
            timestamp: new Date(trans.timestamp).toLocaleString(),
            model: trans.config.selectedModel,
            provider: trans.config.provider
        }
    }
}

File: lib/services/translation_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/app_models.dart';

/// 翻译服务，处理文本翻译逻辑
class TranslationService {
  TranslationResult? _currentTranslation;
  bool _isTranslating = false;

  TranslationResult? get currentTranslation => _currentTranslation;
  bool get isTranslating => _isTranslating;

  /// 翻译文本到多个目标语言
  Future<TranslationResult> translateText(
    String inputText,
    List<String> targetLanguages,
    LLMConfig config,
  ) async {
    if (inputText.trim().isEmpty) {
      throw Exception('Please enter text to translate');
    }

    if (targetLanguages.isEmpty) {
      throw Exception('Please select at least one target language');
    }

    if (config.serverUrl.isEmpty || config.selectedModel.isEmpty) {
      throw Exception('Please configure LLM settings first');
    }

    _isTranslating = true;
    try {
      print('Starting translation request:');
      print('Text: $inputText');
      print('Target languages: $targetLanguages');
      print('Provider: ${config.provider}');
      print('Server URL: ${config.serverUrl}');
      print('Model: ${config.selectedModel}');

      final request = TranslationRequest(
        text: inputText.trim(),
        targetLanguages: targetLanguages,
        provider: config.provider,
        serverUrl: config.serverUrl,
        modelName: config.selectedModel,
        apiKey: config.apiKey,
      );

      final translations = <String, String>{};

      // 逐个语言进行翻译
      for (int i = 0; i < targetLanguages.length; i++) {
        final language = targetLanguages[i];
        print('Translating to $language (${i + 1}/${targetLanguages.length})...');

        String translation;
        switch (config.provider) {
          case 'ollama':
            translation = await _callOllamaTranslate(
              request.text,
              language,
              config.serverUrl,
              config.selectedModel,
            );
            break;
          case 'lmstudio':
            translation = await _callLMStudioTranslate(
              request.text,
              language,
              config.serverUrl,
              config.selectedModel,
              config.apiKey,
            );
            break;
          default:
            throw Exception('Unsupported provider: ${config.provider}');
        }

        translations[language] = translation;
        print('Successfully translated to $language: $translation');
      }

      _currentTranslation = TranslationResult(
        original: inputText.trim(),
        translations: translations,
        timestamp: DateTime.now(),
        languages: targetLanguages,
        config: config,
      );

      print('Translation completed successfully');
      return _currentTranslation!;
    } catch (error) {
      print('Translation failed: $error');
      // 提供更具体的错误信息
      final errorMsg = error.toString();
      if (errorMsg.contains('Connection refused') || errorMsg.contains('network')) {
        throw Exception(
            'Cannot connect to ${config.provider} server at ${config.serverUrl}. Please check if the server is running and accessible.');
      } else if (errorMsg.contains('Unauthorized') || errorMsg.contains('401')) {
        throw Exception('Authentication failed. Please check your API key.');
      } else if (errorMsg.contains('Not found') || errorMsg.contains('404')) {
        throw Exception(
            'Model "${config.selectedModel}" not found. Please check if the model is available.');
      } else if (errorMsg.contains('timeout')) {
        throw Exception('Translation request timed out. The server may be overloaded.');
      } else {
        rethrow;
      }
    } finally {
      _isTranslating = false;
    }
  }

  /// 调用 Ollama API 进行翻译
  Future<String> _callOllamaTranslate(
    String text,
    String targetLanguage,
    String serverUrl,
    String model,
  ) async {
    final url = Uri.parse('$serverUrl/api/generate');
    
    final prompt = '''
Translate the following text to $targetLanguage. Only return the translation, nothing else.

$text''';

    final response = await http.post(
      url,
      headers: {
        'Content-Type': 'application/json',
      },
      body: jsonEncode({
        'model': model,
        'prompt': prompt,
        'stream': false,
        'options': {
          'temperature': 0.1,
          'num_ctx': 4096,
        },
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final translatedText = data['response']?.toString().trim() ?? '';
      
      if (translatedText.isEmpty) {
        throw Exception('Empty translation response from Ollama');
      }
      
      return _cleanTranslationResult(translatedText);
    } else {
      throw Exception('Ollama API request failed: ${response.statusCode} - ${response.body}');
    }
  }

  /// 调用 LM Studio API 进行翻译
  Future<String> _callLMStudioTranslate(
    String text,
    String targetLanguage,
    String serverUrl,
    String model,
    String? apiKey,
  ) async {
    final url = Uri.parse('$serverUrl/v1/chat/completions');
    
    final headers = {
      'Content-Type': 'application/json',
    };
    
    if (apiKey != null && apiKey.isNotEmpty) {
      headers['Authorization'] = 'Bearer $apiKey';
    }

    final response = await http.post(
      url,
      headers: headers,
      body: jsonEncode({
        'model': model,
        'messages': [
          {
            'role': 'system',
            'content': 'You are a professional translator. Only return the translation, nothing else.',
          },
          {
            'role': 'user',
            'content': 'Translate to $targetLanguage:\n\n$text',
          },
        ],
        'temperature': 0.1,
        'max_tokens': 4096,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final choices = data['choices'] as List?;
      
      if (choices != null && choices.isNotEmpty) {
        final message = choices[0]['message'];
        final translatedText = message['content']?.toString().trim() ?? '';
        
        if (translatedText.isEmpty) {
          throw Exception('Empty translation response from LM Studio');
        }
        
        return _cleanTranslationResult(translatedText);
      } else {
        throw Exception('No translation choices returned from LM Studio');
      }
    } else {
      throw Exception('LM Studio API request failed: ${response.statusCode} - ${response.body}');
    }
  }

  /// 清理翻译结果，移除不必要的前缀和后缀
  String _cleanTranslationResult(String result) {
    // 简化处理，只移除明显的前缀和多余空格
    String cleaned = result.trim();
    
    // 移除常见的翻译前缀
    final prefixes = [
      'Translation:',
      'Translated text:',
      'Here is the translation:',
      'The translation is:',
      'Answer:',
      'Response:',
    ];
    
    for (final prefix in prefixes) {
      if (cleaned.toLowerCase().startsWith(prefix.toLowerCase())) {
        cleaned = cleaned.substring(prefix.length).trim();
        break;
      }
    }

    // 移除前后的引号（但要小心处理阿拉伯语等复杂文本）
    if (cleaned.length > 2) {
      if ((cleaned.startsWith('"') && cleaned.endsWith('"')) ||
          (cleaned.startsWith("'") && cleaned.endsWith("'"))) {
        cleaned = cleaned.substring(1, cleaned.length - 1).trim();
      }
    }

    return cleaned;
  }

  /// 获取当前翻译结果
  TranslationResult? getCurrentTranslation() {
    return _currentTranslation;
  }

  /// 清除当前翻译
  void clearTranslation() {
    _currentTranslation = null;
  }

  /// 获取翻译状态
  Map<String, dynamic> getTranslationState() {
    return {
      'isTranslating': _isTranslating,
      'hasTranslation': _currentTranslation != null,
      'currentTranslation': _currentTranslation,
    };
  }

  /// 格式化翻译结果用于显示
  Map<String, dynamic>? formatForDisplay([TranslationResult? translation]) {
    final trans = translation ?? _currentTranslation;
    if (trans == null) return null;

    return {
      'original': trans.original,
      'translations': trans.translations,
      'languages': trans.languages,
      'timestamp': trans.timestamp.toLocal().toString(),
      'model': trans.config.selectedModel,
      'provider': trans.config.provider,
    };
  }
}
